### PR TITLE
allow match in consts

### PR DESCRIPTION
**Pull request by llogiq**
A followup to  #6144, this extends our const handling to allow for `match`. Unsure how we'd test it, though, and we may want to extend it so we can add bindings to expressions *in* matches. Even so, I consider it a win.

r? @ebroto 

---

changelog: none

### CODE REVIEW

**Comment by llogiq** [comment]
Errors seem unrelated, and I cannot reproduce them.

**Comment by ebroto** [comment]
Sorry for the delay in getting this reviewed, I'm really short on time lately. This will be the next PR I'll review when I get some free time.

**Comment by llogiq** [comment]
No worries. I'm also just trying to catch up while following RustFest global...

File: clippy_lints/src/consts.rs [hunk 1]
```diff
@@ -277,11 +280,88 @@ impl<'a, 'tcx> ConstEvalLateContext<'a, 'tcx> {
             },
             ExprKind::Index(ref arr, ref index) => self.index(arr, index),
             ExprKind::AddrOf(_, _, ref inner) => self.expr(inner).map(|r| Constant::Ref(Box::new(r))),
+            ExprKind::Match(ref matchee, &ref arms, _) => {
+                self.expr(matchee).and_then(|m| {
+                    let outer_bindings = self.match_bindings.len();
+                    for arm in arms {
+                        self.match_bindings.truncate(outer_bindings);
```
Comment (ebroto) [inline clippy_lints/src/consts.rs:287]: I think this could be outside of the loop as `outer_bindings` will not change

**Review by ebroto — CHANGES_REQUESTED** [review]
I think we should add tests covering each new possible case in an arbitrary lint that uses the `constant` function.

**Comment by giraffate** [comment]
ping from triage @llogiq. There seem some fixes left to be done. Could you have any update on this?

**Comment by llogiq** [comment]
Yes, but I'll need some more time. If you want you can close and I can reopen when I get around to it.

**Comment by ebroto** [comment]
>Yes, but I'll need some more time. If you want you can close and I can reopen when I get around to it.

I know my requested changes are kind of problematic, sorry about that! I just want to make sure we cover those new paths to avoid triggering bugs that are hard to find because we don't test them.

Feel free to close the PR if you don't find time to adress them currently.

**Comment by ebroto** [comment]
I'm going to close this for now to clean up a bit the PR queue

**Comment by flip1995** [comment]
@llogiq Can you move this branch to a fork, please? This is the only remaining branch, besides the necessary branches for CI, release, and deployment. cc #4745

**Comment by flip1995** [comment]
I pushed this branch to my fork: https://github.com/flip1995/rust-clippy/tree/const-match

I now delete the branch from this repo. We now only have branches necessary for CI, release and deployment in this repo.

**Pull request closed without merging**